Properly support task batches in ReservedStateUpdateTaskExecutor

prdoyle opened the pull request:
`ReservedStateUpdateTaskExecutor` had been extending `SimpleBatchedExecutor` and looping through every state update. Not only was this unnecessary, since only the final state matters; it was also leading to `NullPointerException` in certain cases.

Here, we simplify the processing to execute only the last task.

prdoyle committed:
Unit test ReservedStateUpdateTaskExecutor only running the last task

prdoyle committed:
Fix ReservedStateUpdateTaskExecutor to run only the last task

prdoyle committed:
Spotless

DaveCTurner commented on server/src/main/java/org/elasticsearch/reservedstate/service/ReservedStateUpdateTaskExecutor.java:
```diff
@@ -34,17 +33,27 @@ public ReservedStateUpdateTaskExecutor(RerouteService rerouteService) {
+        try (var ignored = taskContext.captureResponseHeaders()) {
+            var task = taskContext.getTask();
+            clusterState = task.execute(clusterState);
+            taskContext.success(() -> task.listener().onResponse(ActionResponse.Empty.INSTANCE));
```
You have to mark all the other tasks as having succeeded too. There's an assertion to check this in the `MasterService`, so I guess we're missing a test that runs this executor in a sufficiently realistic setup.

prdoyle replied:
(Indeed. I'm not a huge fan of mocks TBH because it's so hard to tell what you're actually testing.</rant>)

DaveCTurner replied:
Yes, I think you could perhaps test this with an `ESSingleNodeTestCase` to get something more realistic.

DaveCTurner replied:
Also it'd be good to have a test that tries to create the same repository twice in a batch in order to demonstrate the NPE is fixed.

n1v0lg replied:
Just thinking out loud on how this might look in an IT: 

In `RepositoriesFileSettingsIT`, we could get a  `FileSettingsService` instance via `internalCluster().getInstance(FileSettingsService.class, masterNode)` and manually invoke `processFileOnServiceStart()` multiple times concurrently. That would result in multiple concurrent and identical state update tasks, though I'm not sure it would suffice to guarantee that they'd all be part of the same batch. 

Based on the `MasterService#createTaskQueue` Javadoc, "Tasks submitted to the same queue (while the master service is otherwise busy) will be batched together into a single cluster state update." -- I'm not sure if there's a simpler way in tests to force batching, aside from ensuring the master is busy with another task when we call `processFileOnServiceStart`.

DaveCTurner replied:
In an `ESIntegTestCase` or an `ESSingleNodeTestCase` you can force batching by first blocking the master service and then doing whatever is needed to submit the tasks you want to be batched, then unblocking the master service again:

```java
        masterNodeClusterService.createTaskQueue("block", Priority.NORMAL, batchExecutionContext -> {
            barrier.await(10, TimeUnit.SECONDS);
            barrier.await(10, TimeUnit.SECONDS);
            batchExecutionContext.taskContexts().forEach(c -> c.success(() -> {}));
            return batchExecutionContext.initialState();
        }).submitTask("block", e -> { assert false : e; }, null);
```

prdoyle replied:
Hi @DaveCTurner - I've finally come back to this. I'm trying to understand your advice here. Looks like the code above is similar to [this code](https://github.com/elastic/elasticsearch/blob/7a98e31f9db4e7155eecc3563284640ea8b5dbf1/server/src/internalClusterTest/java/org/elasticsearch/action/support/AutoCreateIndexIT.java#L30-L35):

```
masterNodeClusterService.createTaskQueue("block", Priority.NORMAL, batchExecutionContext -> {
    safeAwait(barrier);
    safeAwait(barrier);
    batchExecutionContext.taskContexts().forEach(c -> c.success(() -> {}));
    return batchExecutionContext.initialState();
}).submitTask("block", ESTestCase::fail, null);
```

This seems to be a trick to block `masterNodeClusterService` until the other test code encounters the barrier twice. You put whatever code you want to run during the block between the two barriers and it's guaranteed to run after this lamba _begins_ but before it does the `forEach`.

As far as I can tell, the rest is just "legalese" required to make this trick work, and I can pretty much ignore it.

Am I close?

prdoyle replied:
Ok new question... I'm not sure how to trigger this condition in the integration test. I tried bad JSON but that fails in a different way: it never seems to make it to `ReservedStateUpdateTaskExecutor` because it dies earlier, parsing the JSON.

I wonder how I can feed this valid JSON that parses, but then somehow the `ReservedStateUpdateTask` still fails? 🤔

prdoyle replied:
Welp I guess I can at least test the happy path. Submit N  changes and verify that the last one takes effect.

DaveCTurner commented on server/src/main/java/org/elasticsearch/reservedstate/service/ReservedStateUpdateTaskExecutor.java:
```diff
@@ -34,17 +33,27 @@ public ReservedStateUpdateTaskExecutor(RerouteService rerouteService) {
+            clusterState = task.execute(clusterState);
+            taskContext.success(() -> task.listener().onResponse(ActionResponse.Empty.INSTANCE));
+        } catch (Exception e) {
+            taskContext.onFailure(e);
```
If the last task fails, and there are earlier tasks in the list, should we try them instead of completely giving up?

prdoyle replied:
Is it safe to try multiple tasks? I thought the motivation for this PR was to avoid doing so.

DaveCTurner replied:
If the task failed then the cluster state remains unchanged, so yes in that case it's safe to try a different one.

prdoyle replied:
So we're effectively trying them in reverse until one succeeds.

Makes sense I guess, though it seems a little odd. I wonder if fixing it so they can apply in order would be more going "with the grain".

DaveCTurner replied:
That's also a possibility but to achieve that you'd have to go through all the `ClusterStateTaskExecutor` implementations to which the reserved-state subsystem delegates checking for batchability problems.

n1v0lg replied:
Every failed task will also kick off a separate task to update error state in reserved cluster state metadata, and these error state updates are also subject to the same version checks. So if we have three tasks with versions (`v3` highest, `v1` lowest):

`(t3, v3), (t2, v2), (t1, v1)` 

We'd apply `t3` first. Suppose it fails. Then error state is marked with version `v3`. When we process `t2`, if it succeeds, we clear the error regardless of version, so we're fine. But if it fails, I think we will skip updating the error state because `v3` > `v2`.  

The code for error handling is in several places, but see [here](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/reservedstate/service/ReservedClusterStateService.java#L297C10-L304) for example

I wonder if we should simply try to apply the highest version task for now, then follow up with potential improvements (like trying older tasks, or instead trying the queued tasks in reverse order).

DaveCTurner replied:
That's not my understanding. If we fail to update state `t3` then we spawn a _separate_ task (or batch thereof) that will run after the current batch is complete, so when we process `t2` there's no error state to clear.

n1v0lg replied:
You are very likely right, but let me just write out what the flow looks like:

1. we submit the state update task [here](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/reservedstate/service/ReservedClusterStateService.java#L237)
2. the update task takes an error-handler (`ReservedClusterStateService.this::updateErrorState`) 
3. when invoked, that error handlers [submits an error state update task](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/reservedstate/service/ReservedClusterStateService.java#L297-L305) (to a separate executor, with a separate queue)
4. Inside the cluster-state update task `execute` method, if we run into errors, we trigger [the error handler](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/reservedstate/service/ReservedStateUpdateTask.java#L144) and throw an exception

Given that error update tasks go to a separate executor and task queue, is it still true that they're guaranteed to run _after_ we've processed tasks in the non-error state update queue?

That is, can't it happen that we finish task `t3`, which kicks of error task `t3_error`, which then completes before we get to task `t2`? I'm probably just misunderstanding what order guarantees we have for cluster-state update tasks stored in separate queues and handled by separate executors.

n1v0lg replied:
> Yes. We're already processing a fixed batch of tasks here, there's no way to add more tasks to it. Even if the error-update tasks went to the same executor, they'd be added to a new batch of tasks, allowing us to complete the processing of the current batch of tasks undisturbed.

@DaveCTurner thanks for clarifying! The piece I was missing conceptually is that the execution of batches is also ordered so while we're processing one batch, we won't start processing another batch in parallel. 

@prdoyle all good to ignore my comments 👍

DaveCTurner reviewed (commented):
That's the basic idea yeah, I left a handful of suggestions tho.

prdoyle commented on server/src/main/java/org/elasticsearch/reservedstate/service/ReservedStateUpdateTaskExecutor.java:
```diff
@@ -34,17 +33,27 @@ public ReservedStateUpdateTaskExecutor(RerouteService rerouteService) {
+            return initState;
+        }
+        // Only the last update is relevant; the others can be skipped
+        var taskContext = taskContexts.get(taskContexts.size() - 1);
```
Oh... is there some other way to define the "order" of tasks?

DaveCTurner replied:
I'm not sure, but I understood that file-based settings have some notion of version (carried in the file) to ensure we're not overriding fresh settings with stale ones. I believe it relates to `task.stateChunk.metadata().version()`, although I haven't tracked down the details.

n1v0lg replied:
Correct, `task.stateChunk.metadata().version()` will give you the version of the file-settings file this update task is for. This is then [compared against](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/reservedstate/service/ReservedStateUpdateTask.java#L94-L96) the last-processed version stored in metadata.

We should pick the update task with the highest version as @DaveCTurner suggests, unless someone from core infra confirms that the tasks are already guaranteed to be "well-ordered".

DaveCTurner commented:
> Given that error update tasks go to a separate executor and task queue, is it still true that they're guaranteed to run after we've processed tasks in the non-error state update queue?

Yes. We're already processing a fixed batch of tasks here, there's no way to add more tasks to it. Even if the error-update tasks went to the same executor, they'd be added to a new batch of tasks, allowing us to complete the processing of the current batch of tasks undisturbed.

prdoyle commented:
Alright I think I can do all of the above, though I'd gratefully accept some help with the tests. 😅

prdoyle committed:
Use penultimate update if the last one failed

prdoyle committed:
Merge branch 'main' into es-9360-2

prdoyle committed:
Fix test method name to match what it does

prdoyle committed:
IT for testLastSettingsInBatchApplied

DaveCTurner commented on server/src/internalClusterTest/java/org/elasticsearch/reservedstate/service/FileSettingsServiceIT.java:
```diff
@@ -381,6 +384,40 @@ public void testErrorSaved() throws Exception {
+        }).submitTask("block", ESTestCase::fail, null);
+
+        safeAwait(barrier);
+        writeJSONFile(masterNode, testJSON, versionCounter, logger);      // Valid but skipped
+        writeJSONFile(masterNode, testJSON43mb, versionCounter, logger);  // The last valid setting
```
Could we sometimes try 3, I have a hunch that might cover some logic that the 2 case doesn't.

Could we give them different `version` values? And not always apply them in ascending order?

Could we wait for the master to pick up one file before writing the next?

prdoyle replied:
Good ideas!

Waiting for the master to pick up one file before writing the next would exercise existing functionality, rather than my changes, but it's obviously an important case to test.

prdoyle replied:
What is the expected behaviour when the versions are not ascending?

Should `ReservedStateUpdateTaskExecutor` be attempting them in descending version order rather than just iterating backward through the batch list?

DaveCTurner replied:
The guiding principle is that the expected outcome should be the same whether the tasks are executed in a single batch or one-at-a-time. So yes AIUI I think we should attempt them in descending version order.

prdoyle replied:
Ah of course, that makes perfect sense. I'll make that change.

prdoyle replied:
(Again, it seems like we're going to a lot of trouble to simulate running them in order, instead of actually running them in order.)

DaveCTurner replied:
You have to call `success` (or `onFailure`) to record the outcome of the task, but that doesn't mean you have to actually do anything else with the task. According to the guiding principle, skipping a task because we processed a newer one counts as success right?

prdoyle replied:
Alright makes sense. So perhaps:

1. Scan the tasks to identify the one that would have run last, had the tasks all been run individually instead of in a batch, taking into consideration task ordering, version numbers, and the `ReservedStateVersionCheck` mode.
2. Try that task. If it succeeds, call `success` on that one and all prior ones, and possibly on all following ones?
4. If it fails, call `onFailure`, remove it from the candidates list, and loop back to #1.

DaveCTurner replied:
Yeah that sounds right although maybe it'd be simpler just to sort the list of tasks in the right order ?

prdoyle replied:
I was afraid I couldn't determine the right order before I know whether the first fails. But now that I think about it, if the first task succeeds, the rest are irrelevant, so I could sort them on the assumption that each one fails.

DaveCTurner commented:
That's why I [suggested](https://github.com/elastic/elasticsearch/pull/116353#discussion_r1861726455) waiting for the master to pick up each change and enqueue the relevant task before overwriting the file again.

prdoyle commented:
Oh I see what you meant. I thought you meant they'd end up in different batches.

prdoyle committed:
Merge remote-tracking branch 'upstream/main' into es-9360-2

prdoyle committed:
Consider version numbers in upsdate batches

prdoyle committed:
Track and assert on task completion

prdoyle committed:
Assert each task's state

prdoyle committed:
Use sorting.

This will be O(n log n) for long lists, while the original way will be O(mn)
where m is the number of failing tasks.

None of this matters much for performance because n is expected to be small;
but it does simplify the code and make it smell less imperative.

prdoyle committed:
Merge remote-tracking branch 'upstream/main' into es-9360-2

prdoyle commented:
Ok, I have implemented the precedence logic we discussed, used sorting over the tasks, and added unit tests to ensure all the tasks are marked completed.

All that remains, I think, is to change the IT so it forces multiple file changes to appear in the same batch, in order to prove the original NPE (which motivated this PR) is gone.

@DaveCTurner - do you happen to have any tips for how to achieve that? I assume I can use another barrier in a strategic spot to make the file watcher and unit test wait for each other?

elasticsearchmachine commented:
Pinging @elastic/es-core-infra (Team:Core/Infra)

prdoyle committed:
Spotless

DaveCTurner commented:
> do you happen to have any tips for how to achieve that

Assuming there's no easy way to wait for the file watcher to run, I think I'd busy-wait, checking `org.elasticsearch.cluster.service.MasterService#pendingTasks` repeatedly until I see the first task, then update the file, then busy-wait on `org.elasticsearch.cluster.service.MasterService#pendingTasks` again until I see the second task, and so on.

prdoyle committed:
Merge branch 'main' into es-9360-2

prdoyle committed:
Fixes and comments

prdoyle commented:
I might need a way to wait for the watcher anyway. The way I've written the IT tests, there's no guarantee all my changes have taken effect before the assertions run. It sometimes results in this:

```
FileSettingsServiceIT > testLastSettingsInBatchApplied FAILED
    java.lang.AssertionError: 
    Expected: "43mb"
         but: was "50mb"
        at __randomizedtesting.SeedInfo.seed([76FB2677EDF08BC8:E940E5E7DD70379E]:0)
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:6)
        at org.elasticsearch.test.ESTestCase.assertThat(ESTestCase.java:2581)
        at org.elasticsearch.reservedstate.service.FileSettingsServiceIT.assertExpectedRecoveryBytesSettingAndVersion(FileSettingsServiceIT.java:226)
        at org.elasticsearch.reservedstate.service.FileSettingsServiceIT.assertClusterStateSaveOK(FileSettingsServiceIT.java:218)
        at org.elasticsearch.reservedstate.service.FileSettingsServiceIT.testLastSettingsInBatchApplied(FileSettingsServiceIT.java:418)
```

prdoyle commented:
Trying to wait for the watcher ended up deadlocking due to [this line](https://github.com/elastic/elasticsearch/blob/00a1222f10a6bc605f67aee67d4053c5ba0557e8/server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java#L173), where the watcher waits for the submitted task to complete before processing the next file.

This makes me wonder how it's even _possible_ to have a batch containing more than one file update. It seems like `FileSettingsService` blocks the watcher thread until the file processing is finished.

prdoyle commented:
@DaveCTurner - can I get a sanity check here? I think this whole PR might be unnecessary, and the problem could be deeper.

File changes are processed on the `watcherThread`, of which there can be at most one per node. (**Caveat**: more on this in a moment.) When the file changes, that calls (indirectly) `FileSettingsService.processFileChanges`, which:
1. makes its way to `ReservedClusterStateService.process`, which creates the update task and submits it to the `updateTaskQueue`; and then
2. calls `completion.get()` to wait for the task to finish.

This makes it impossible for a second file update task to be created until the first is finished, which implies that it's impossible for two update tasks to be in the same batch.

This would imply that what we have observed is impossible, _unless more than one thread submits update tasks_.

I could see this happening in two scenarios:

1. `processFileOnServiceStart` races with `processSettingsAndNotifyListeners`, or
2. we accidentally spawn a second watcher thread.

It seems to be that these are both undesirable, and if we fix those, we no longer need to worry about multiple updates per batch.

The first could be fixed with appropriate synchronization.

The second could be fixed by using a singleton executor instead of directly using `new Thread(...).start()`.

What do you think?

DaveCTurner commented:
Ah interesting, I hadn't looked in much detail at the calling code, I didn't realise it was doing stuff with bare `Thread` calls and interrupts and so on. It's not really the Elasticsearch way, tho in this case it's only one extra thread so scalability isn't a huge concern.

So yes another possibility would be to ensure that there's only ever at most one of these tasks in the master queue at once.

Note that today's implementation doesn't guarantee this, because the wait for the task to complete may be interrupted if the node stops being master. It's then possible for the node to become master again straight away, start up a new watcher thread and enqueue another task all while the old task is still in the queue. Instead we must always wait for the task's listener to complete.

prdoyle commented:
Thanks @DaveCTurner. What would be the Elasticsearch Way to implement this? Is it a good idea to change it?

On other projects, I've used `scheduleWithFixedDelay` on a singleton `ScheduledExecutorService`, which has the side benefit that if the thread exits for any reason, it will wait a bit and then restart it automatically. I suspect we'd prefer something more `ThreadPool` oriented?

prdoyle commented:
Ok I think I'd rather actually _use_ a bina fide dedicated thread pool of size 1 then, so it doesn't accidentally end up with two threads causing otherwise impossible race conditions.

rjernst commented:
I'm still not sure we should use ThreadPool. That comes with baggage (eg, size configurable by user through elasticsearch.yml). This is an implementation detail. We should make it robust, but IMO we should do so without ThreadPool.

prdoyle commented:
Ok makes sense. I really meant lowercase "thread pool" as a general concept. So if you prefer to avoid `ThreadPool` here, it could just be some sort of `Executor`.

DaveCTurner commented:
Yes, we do create a dedicated executor in other places - see e.g. `org.elasticsearch.cluster.service.MasterService#createThreadPoolExecutor`.

prdoyle commented:
@DaveCTurner - what if I change `ReservedClusterStateService` to create a new task queue for each `ReservedStateUpdateTask` instead of using the same long-lived queue? That would force each task into its own batch I think?

prdoyle commented:
New PR #118112 that aims to prevent batching rather than cope with it. FYI @DaveCTurner @rjernst

prdoyle commented:
Superseded by #118112
